common/dev/ltc4282.c

```diff
@@ -1,200 +1,283 @@
+			return SENSOR_UNSPECIFIED_ERROR;
+		}
+	}
+
+	uint8_t retry = 5;
+	double val = 0;
+	I2C_MSG msg = { 0 };
+
+	sensor_cfg *cfg = &sensor_config[sensor_config_index_map[sensor_num]];
+
+	msg.bus = cfg->port;
+	msg.target_addr = cfg->target_addr;
+	msg.tx_len = 1;
+	msg.data[0] = cfg->offset;
+	msg.rx_len = 2;
+
+	if (cfg->offset != LTC4282_ENERGY_OFFSET) {
+		if (i2c_master_read(&msg, retry) != 0)
+			return SENSOR_FAIL_TO_ACCESS;
+	}
+	// Refer to LTC4282 datasheet page 23.
+	switch (cfg->offset) {
+	case LTC4282_VSENSE_OFFSET:
+		val = (((msg.data[0] << 8) | msg.data[1]) * 0.04 / 65535 / rsense_mohm);
+		break;
+	case LTC4282_POWER_OFFSET:
+		val = (((msg.data[0] << 8) | msg.data[1]) * full_scale_output_voltage * 0.04 *
+		       65536 / 65535 / 65535 / rsense_mohm);
+		break;
+	case LTC4282_VSOURCE_OFFSET:
+		val = (((msg.data[0] << 8) | msg.data[1]) * full_scale_output_voltage / 65535);
+		break;
+	case LTC4282_ENERGY_OFFSET:
+		if (ltc4282_read_ein(&msg, &val, retry) < 0) {
+			return SENSOR_FAIL_TO_ACCESS;
+		}
+		val = (val * 16.64 * 0.04 * 256 / 65535 / 65535 / rsense_mohm);
+		break;
+	default:
+		printf("Invalid sensor 0x%x offset 0x%x\n", sensor_num, cfg->offset);
+		return SENSOR_NOT_FOUND;
+	}
+
+	sensor_val *sval = (sensor_val *)reading;
+	memset(sval, 0, sizeof(*sval));
+
+	sval->integer = (int)val & 0xFFFF;
+	sval->fraction = (val - sval->integer) * 1000;
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t ltc4282_init(uint8_t sensor_num)
+{
+	if (sensor_num > SENSOR_NUM_MAX) {
+		return SENSOR_INIT_UNSPECIFIED_ERROR;
+	}
+
+	I2C_MSG msg;
+	uint8_t retry = 5;
+
+	msg.bus = sensor_config[sensor_config_index_map[sensor_num]].port;
+	msg.target_addr = sensor_config[sensor_config_index_map[sensor_num]].target_addr;
+
+	ltc4282_init_arg *init_args =
+		(ltc4282_init_arg *)sensor_config[sensor_config_index_map[sensor_num]].init_args;
+
+	if (init_args->is_init) {
+		goto exit;
+	}
+
+	if (init_args->is_register_setting_needed == 0x0) {
+		goto init_param;
+	}
+
+	int bit = 0;
+	do {
+		switch (init_args->is_register_setting_needed & (BIT(bit))) {
+		case BIT(0):
+			// Set ILIM_ADJUST register
+			msg.tx_len = 2;
+			msg.rx_len = 0;
+			msg.data[0] = LTC4282_ILIM_ADJUST_OFFSET;
+			msg.data[1] = init_args->ilim_adjust.value;
+			if (i2c_master_write(&msg, retry) != 0) {
+				printf("Failed to set LTC4282 register(0x%x)\n", msg.data[0]);
+				goto exit;
+			}
+			break;
+		default:
+			break;
+		}
+		bit++;
+	} while (bit < 8);
+
+init_param:
+	// Read ILIM_ADJUST register value
+	msg.tx_len = 1;
+	msg.rx_len = 1;
+	msg.data[0] = LTC4282_ILIM_ADJUST_OFFSET;
+	if (i2c_master_write(&msg, retry) != 0) {
+		printf("Failed to read LTC4282 register(0x%x)\n", msg.data[0]);
+		init_args->is_init = false;
+		goto exit;
+	}
+	init_args->ilim_adjust.value = msg.data[0];
+
+	init_args->is_init = true;
+
+exit:
+	sensor_config[sensor_config_index_map[sensor_num]].read = ltc4282_read;
+	return SENSOR_INIT_SUCCESS;
```

Comment:
Should this return SENSOR_INIT_SUCCESS since it reached the exit from a failed i2c_master_write in line 271?

Reply:
The code is forced updated.
Please review it again.
Thanks.